cirq/sim/density_matrix_simulator.py

```diff
@@ -0,0 +1,425 @@
+        measurements = {}  # type: Dict[str, List[np.ndarray]]
+        # TODO: optimize for all terminal measurements.
+        for _ in range(repetitions):
+            all_step_results = self._base_iterator(
+                resolved_circuit,
+                qubit_order=ops.QubitOrder.DEFAULT,
+                initial_state=0,
+                perform_measurements=True)
+            for step_result in all_step_results:
+                for k, v in step_result.measurements.items():
+                    if not k in measurements:
+                        measurements[k] = []
+                    measurements[k].append(np.array(v, dtype=bool))
+        return {k: np.array(v) for k, v in measurements.items()}
+
+
+    def _simulator_iterator(self,
+            circuit: circuits.Circuit,
+            param_resolver: study.ParamResolver,
+            qubit_order: ops.QubitOrderOrList,
+            initial_state: Union[int, np.ndarray]) -> Iterator:
+        """See definition in `cirq.SimulatesIntermediateState`.
+
+        If the initial state is an int, the state is set to the computational
+        basis state corresponding to this state. Otherwise  if the initial
+        state is a np.ndarray it is the full initial state, either a pure state
+        or the full density matrix.  If it is the pure state it must be the
+        correct size, be normalized (an L2 norm of 1), and be safely castable
+        to an appropriate dtype for the simulator.  If it is a mixed state
+        it must be correctly sized and positive semidefinite with trace one.
+        """
+        param_resolver = param_resolver or study.ParamResolver({})
+        resolved_circuit = protocols.resolve_parameters(circuit, param_resolver)
+        actual_initial_state = 0 if initial_state is None else initial_state
+        return self._base_iterator(resolved_circuit,
+                                   qubit_order,
+                                   actual_initial_state)
+
+    def _base_iterator(
+        self,
+        circuit: circuits.Circuit,
+        qubit_order: ops.QubitOrderOrList,
+        initial_state: Union[int, np.ndarray],
+        perform_measurements: bool=True,
+    ) -> Iterator:
+        qubits = ops.QubitOrder.as_qubit_order(qubit_order).order_for(
+            circuit.all_qubits())
+        num_qubits = len(qubits)
+        qubit_map = {q: i for i, q in enumerate(qubits)}
+        matrix = density_matrix_utils.to_valid_density_matrix(
+            initial_state, num_qubits, self._dtype)
+        if len(circuit) == 0:
+            yield DensityMatrixStepResult(matrix, {}, qubit_map, self._dtype)
+
+        def on_stuck(bad_op: ops.Operation):
+            return TypeError(
+                "Can't simulate operations that don't implement "
+                "SupportsUnitary, SupportsApplyUnitary, SupportsMixture, "
+                "SupportsChannel or is a measurement: {!r}".format(bad_op))
```

Comment:
Future: we may want some way to include the concept of measurement into our protocols, so people can define e.g. measurements by different observables.

Reply:
Yep, this is #1357 I think.